cezo_fl/util/prepare_settings.py

```diff
@@ -0,0 +1,234 @@
+from typing import Literal, Callable
+import torch
+import torch.nn as nn
+
+from functools import partial
+from peft import LoraConfig, get_peft_model
+from transformers import AutoModelForCausalLM
+
+
+from cezo_fl.util import model_helpers
+from cezo_fl.util.model_helpers import AllModel
+from cezo_fl.models.cnn_fashion import CNN_FMNIST
+from cezo_fl.models.cnn_mnist import CNN_MNIST
+from cezo_fl.models.lenet import LeNet
+from cezo_fl.models.lstm import CharLSTM
+from cezo_fl.random_gradient_estimator import RandomGradientEstimator as RGE
+from cezo_fl.util.language_utils import (
+    LM_TEMPLATE_MAP,
+    SUPPORTED_LLM,
+    get_lm_loss,
+    get_hf_tokenizer,
+)
+from cezo_fl.util.metrics import accuracy
+
+from dataclasses import dataclass
+
+
+def get_torch_dtype(args_model_dtype: Literal["float32", "float16", "bfloat16"]) -> torch.dtype:
+    return {
+        "float32": torch.float32,
+        "float16": torch.float16,
+        "bfloat16": torch.bfloat16,
+    }[args_model_dtype]
+
+
+def get_random_gradient_estimator(args, model, device):
+    if args.grad_estimate_method in ["rge-central", "rge-forward"]:
+        return RGE(
+            parameters=model_helpers.get_trainable_model_parameters(model),
+            mu=args.mu,
+            num_pert=args.num_pert,
+            grad_estimate_method=args.grad_estimate_method,
+            device=device,
+            torch_dtype=get_torch_dtype(args.model_dtype),
+            # To save memory consumption, we have to use parameter-wise perturb + no_optim together.
+            sgd_only_no_optim=args.no_optim,
+            paramwise_perturb=args.no_optim,
+        )
+    else:
+        raise Exception(f"Grad estimate method {args.grad_estimate_method} not supported")
+
+
+def get_model_and_optimizer(
+    dataset: str,
+    model_dtype: Literal["float32", "float16", "bfloat16"],
+    lr: float,
+    momentum: float,
+    large_model: str | None,
+    lora: bool = False,
+    lora_alpha: int = 16,
+    lora_r: int = 8,
+):
+    torch_dtype = get_torch_dtype(model_dtype)
+    model: AllModel
+    if dataset == "mnist":
+        model = CNN_MNIST().to(torch_dtype)
+        optimizer = torch.optim.SGD(
+            model_helpers.get_trainable_model_parameters(model),
+            lr=lr,
+            weight_decay=1e-5,
+            momentum=momentum,
+        )
+
+        # scheduler = torch.optim.lr_scheduler.ExponentialLR(optimizer, gamma=0.8)
```

Comment:
let's delete this. If we need scheduler, we can make another function dedicated for scheduler.